### ISSUE TITLE
Use SHA256/512 for HMAC operations

### ISSUE THREAD

**Issue by Synchro** [laravel/ideas#1303]
The new signed URLs feature in Laravel 5.7 [uses HMAC-SHA256](https://github.com/laravel/framework/blob/3414dcfcbe27cf0f4deee0670f022983e8016392/src/Illuminate/Routing/UrlGenerator.php#L319). Something that's common elsewhere is to use HMAC-SHA512/256 instead; this is simply SHA512 truncated to 256 bits. While it might seem counterintuitive, SHA512 is significantly faster on 64-bit CPUs because SHA256 is only a 32-bit algorithm. SHA256 is faster on short strings of less than 448bits of input length, but SHA512 becomes faster (up to a theoretical maximum of 60% faster), which is long enough to include even URLs of moderate length. There's a nice discussion of their relative performance [here](https://crypto.stackexchange.com/questions/26336/sha512-faster-than-sha256).

Security-wise, while both SHA256 and SHA512 hash functions are vulnerable to length extension attacks, SHA512/256 is not, but the HMAC algorithm protects them against that anyway.

Laravel 5.7 requires a minimum version of PHP 7.1.3, and coincidentally PHP 7.1.0 added SHA512/256 to the `hash` and `hash_hmac` functions, so there is no backward compatibility issue with PHP itself.

For Laravel to switch to this algorithm, signature generation can simply be changed without consequence:

        return $this->route($name, $parameters + [
            'signature' => hash_hmac('sha512/256', $this->route($name, $parameters), $key),
        ]);

Signature validation in [`hasValidSignature`](https://github.com/laravel/framework/blob/3414dcfcbe27cf0f4deee0670f022983e8016392/src/Illuminate/Routing/UrlGenerator.php#L350) would need to change to check signature failures against SHA256 as well to ensure backward compatibility, along the lines of:

        $signature = hash_hmac('sha512/256', $original, call_user_func($this->keyResolver));
        $signatureok = hash_equals($signature, $request->query('signature', ''));
        if (!$signatureok) {
            $signature = hash_hmac('sha256', $original, call_user_func($this->keyResolver));
            $signatureok = hash_equals($signature, $request->query('signature', ''));
        }
        return  $signatureok && ! ($expires && Carbon::now()->getTimestamp() > $expires);

FWIW, libsodium uses HMAC-SHA512/256 for the [`sodium_crypto_auth`](http://php.net/manual/en/function.sodium-crypto-auth.php) function, and is about 10% faster than `hash_hmac` for 256byte strings, though that obviously requires PHP 7.2 or pecl-sodium, and has some minor practicalities that mean it's probably simpler to stick with `hash_hmac`.

**Comment by mfn** [comment]
But I think you failed to mention the BC break within Laravel 5.6 -> 5.7, no? Because this code was introduced in 5.6 AFAICS.

**Comment by Synchro** [comment]
Ah, I'd only seen it in conjunction with signed URLs, which AFAIK are new in 5.7 - is it used elsewhere? Laravel 5.6 docs still says 7.1.3 as minimum PHP version. Either way, the BC fallback in validation should work wherever it's used.

**Comment by aufalmarom** [comment]
how to use or import hash_hmac function in header controller?

**Comment by aufalmarom** [comment]
i wanna generate $key with $key = hash_hmac('sha256,', $url, $params);
but
my problem is = hash_hmac(): Unknown hashing algorithm: sha256,
please help me

**Comment by michaeldyrynda** [comment]
Remove the comma inside the first parameter. `'sha256'`, not `'sha256,'`

**Comment by crynobone** [comment]
Is it really fully backward compatible, as in existing signed URL generated by the app will continue to work if this changes has been made?

**Comment by Synchro** [comment]
Yes, because in my example code, it checks the sha512/256 signature first, and if that fails, it tries the sha256 MAC instead, so it falls back transparently, with a minor performance hit.

**Comment by mubarak23** [comment]
How I can decode hmacsha256 string and get the data out of it.

**Comment by Synchro** [comment]
@mubarak23 You can't – that's most of the point of a hash or a MAC. You can put data in, but you can't get it out, but, given something that *should be* the same as the original data, you can confirm that it *really is* the same by repeating the operation and comparing the results. This is how PHP's standard password hashing works. You can also save the hash or MAC alongside the original data and use it as an identifier without any risk of leaking the original data.
